common/src/main/java/io/pravega/common/tracing/RequestTracker.java

```diff
@@ -0,0 +1,182 @@
+     * Retrieves a RequestTag object formed by a request descriptor and request id pair. In the case of concurrent
+     * requests with the same descriptor,
+     *
+     * @param requestInfo Fields to form the request descriptor.
+     * @return Request descriptor and request id pair embedded in a RequestTag object.
+     */
+    public RequestTag getRequestTagFor(String...requestInfo) {
+        return getRequestTagFor(RequestTracker.buildRequestDescriptor(requestInfo));
+    }
+
+    public RequestTag getRequestTagFor(String requestDescriptor) {
+        Preconditions.checkArgument(requestDescriptor != null, "Attempting to get a null request descriptor.");
+        List<Long> requestIds = ongoingRequests.getIfPresent(requestDescriptor);
+        if (requestIds == null) {
+            log.warn("Attempting to get a non-existing tag: {}.", requestDescriptor);
+            return new RequestTag(requestDescriptor, RequestTag.NON_EXISTENT_ID);
+        } else if (requestIds.size() > 1) {
+            log.warn("{} request ids are associated with same descriptor: {}. The first request id {} is the one that will be propagated, " +
+                "given that other operations will not take effect due to idempotence of traced operations.", requestIds,
+                    requestDescriptor, requestIds.get(0));
+        }
+
+        return new RequestTag(requestDescriptor, requestIds.get(0));
+    }
+
+    /**
+     * Retrieves a request id associated to a request descriptor.
+     *
+     * @param requestInfo Fields to form the request descriptor.
+     * @return Request descriptor and request id pair embedded in a RequestTag object.
+     */
+    public long getRequestIdFor(String...requestInfo) {
+        return getRequestIdFor(RequestTracker.buildRequestDescriptor(requestInfo));
+    }
+
+    public long getRequestIdFor(String requestDescriptor) {
+        return getRequestTagFor(requestDescriptor).getRequestId();
+    }
+
+    /**
+     * Adds a request descriptor and request id pair in the cache. In the case of tracking a request with an existing
+     * descriptor, this method adds the request id to the list associated to the descriptor.
+     *
+     * @param requestTag Request to be cached for further tracing.
+     */
+    public void trackRequest(RequestTag requestTag) {
+        trackRequest(requestTag.getRequestDescriptor(), requestTag.getRequestId());
+    }
+
+    public synchronized void trackRequest(String requestDescriptor, long requestId) {
+        Preconditions.checkArgument(requestDescriptor != null, "Attempting to track a null request descriptor.");
+        List<Long> requestIds = ongoingRequests.getIfPresent(requestDescriptor);
+        if (requestIds == null) {
+            requestIds = new ArrayList<>();
+        }
+
+        requestIds.add(requestId);
+        ongoingRequests.put(requestDescriptor, requestIds);
+        log.debug("Tracking request {} with id {}. Current ongoing requests: {}.", requestDescriptor, requestId,
```

Comment:
This operation is expensive and will be calculated every time this method is invoked, regardless of your debug level.
Please figure out a better way to evaluate it or drop it entirely `ongoingRequests.asMap().values().stream().mapToInt(List::size).sum()`

Reply:
Please fix this below too

Reply:
This was there mainly for debug purposes, I removed that.